lib/render/RayCaster.cpp

```diff
@@ -765,156 +922,210 @@ void RayCaster::_drawVolumeFaces( int            whichPass,
-    glUniform1i( uniformLocation, textureUnit );
+    glUniform1i( uniformLocation,  _frontFaceTexOffset );
 
-    textureUnit = 2;
-    glActiveTexture( GL_TEXTURE0 + textureUnit );
-    glBindTexture( GL_TEXTURE_3D, _volumeTextureId );
+    glActiveTexture( GL_TEXTURE0 + _volumeTexOffset );
+    glBindTexture( GL_TEXTURE_3D,  _volumeTextureId );
     uniformLocation = glGetUniformLocation( _3rdPassShaderId, "volumeTexture" );
-    glUniform1i( uniformLocation, textureUnit );
+    glUniform1i( uniformLocation,  _volumeTexOffset );
 
-    textureUnit = 3;
-    glActiveTexture( GL_TEXTURE0 + textureUnit );
-    glBindTexture( GL_TEXTURE_1D, _colorMapTextureId );
+    glActiveTexture( GL_TEXTURE0 + _colorMapTexOffset );
+    glBindTexture( GL_TEXTURE_1D,  _colorMapTextureId );
     uniformLocation = glGetUniformLocation( _3rdPassShaderId, "colorMapTexture" );
-    glUniform1i( uniformLocation, textureUnit );
+    glUniform1i( uniformLocation,  _colorMapTexOffset );
 
-    uniformLocation = glGetUniformLocation( _3rdPassShaderId, "hasMissingValue" );
-    // If there is missing value, pass in missingValueMaskTexture as well.
-    //   Otherwise, leave it empty.
-    if( _userCoordinates.missingValueMask == nullptr )
-        glUniform1i( uniformLocation, 0 );      // Set to false
-    else
+    glActiveTexture(  GL_TEXTURE0 +     _missingValueTexOffset );
+    glBindTexture(    GL_TEXTURE_3D,    _missingValueTextureId );
+    uniformLocation = glGetUniformLocation( _3rdPassShaderId, "missingValueMaskTexture" );
+    glUniform1i(      uniformLocation,  _missingValueTexOffset );
+
+    if( castingMode == 2 )
     {
-        glUniform1i( uniformLocation, 1 );      // Set to true
-        textureUnit = 4;
-        glActiveTexture( GL_TEXTURE0 + textureUnit );
-        glBindTexture( GL_TEXTURE_3D, _missingValueTextureId );
-        uniformLocation = glGetUniformLocation( _3rdPassShaderId, "missingValueMaskTexture" );
-        glUniform1i( uniformLocation, textureUnit );
+        glActiveTexture(  GL_TEXTURE0 +      _xyCoordsTexOffset );
+        glBindTexture(    GL_TEXTURE_BUFFER, _xyCoordsTextureId );
+        uniformLocation = glGetUniformLocation( _3rdPassShaderId, "xyCoordsTexture" );
+        glUniform1i(      uniformLocation,   _xyCoordsTexOffset );
+
+        glActiveTexture(  GL_TEXTURE0 +      _zCoordsTexOffset );
+        glBindTexture(    GL_TEXTURE_BUFFER, _zCoordsTextureId );
+        uniformLocation = glGetUniformLocation( _3rdPassShaderId, "zCoordsTexture" );
+        glUniform1i(      uniformLocation,   _zCoordsTexOffset );
     }
 }
     
-void RayCaster::_3rdPassSpecialHandling( bool fast )
+void RayCaster::_3rdPassSpecialHandling( bool fast, long castingMode )
 {
     // Left empty intentially.
     // Derived classes feel free to put stuff here.
 }
     
-void RayCaster::_renderTriangleStrips() const
+void RayCaster::_renderTriangleStrips( int whichPass, long castingMode ) const
```

Comment:
This method is too long with lots of repetition. Please use helper functions and reduce the repetitive code.

Reply:
This method seems to contain repetitions, but each face has different indices, and very hard to parameterize and deploy helper functions. Further, this method contains many OpenGL function calls that have very tricky arguments. Currently this function is fairly optimized and works well, I'd suggest we leave it alone at this point. What do you think?

Reply:
suggested changes to _renderTriangleStrips()

1. Better document rendering loops
2. Clearer variable names
3. Move opengl setup code at beginning of each loop into a single function, parameterized as needed.

Reply:
Fixed in ddf07345